keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,117 @@
+  - [Proposal](#proposal)
+    - [Implementation Details](#implementation-details)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+  - [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Slow starting containers are difficult to address with the current status of health probes: you either kill them before being up, or you have the risk of leaving dead-locked for a very long time.
+
+I am suggesting to add a numerical option `maxInitialFailureCount` to probes to allow a greater number of failures during the initial start of the container before taking action, while keeping `failureThreshold` at a minimum to restart dead-locked containers after an acceptable delay.
+
+## Motivation
+
+The main problem with slow starting containers is that you have to leave them enough time to start before having `livenessProbe` fail `failureThreshold` times (otherwise `kubelet` kills them before they have a chance to be up).
+This is even worse when load on the nodes can increase this long startup time since you have to take into account the worse case scenario (let's call it **T**).
+
+You can handle that today with various strategies:
+
+- set `initialDelaySeconds` > **T**: the `livenessProbe` isn't run for **T** time (so you're blind in the monitoring), you have to wait at least **T** until `kubelet` kills your container after a deadlock - this could only happen at the beginning of the container lifetime
+- set `failureThreshold` * `periodSeconds` > **T**: the `livenessProbe` is run every `periodSeconds` (so you can monitor the probe status), **but you still** have to wait at least **T** until `kubelet` kills your container after a deadlock - this could happen at any time
+
+### Goals
+
+Add an elegant and self-explanatory solution to running slow starting containers in Kubernetes with health probes enabled.
+
+By using `maxInitialFailureCount` * `periodSeconds` > **T** you have the following benefits:
```

Comment:
This belongs under the proposal.  The proposal is specifically for `maxInitialFailureCount`, so start by noting that, then describe how it works.  The description of the mechanism needs to be more specific, that failures up to a point (`maxInitialFailureCount`) are ignored until the container has successfully started.  Once a probe has ever succeeded, the container is considered to have been initialized, and the existing criteria are used to determine failure.

For example, if `periodSeconds` is 10, `maxInitialFailureCount` is 20, and `failureThreshold` is 3:

- The kubelet will allow the kubelet 200 seconds to start (20 probes, spaced 10 seconds apart).
- If a probe succeeds at any time during that interval, the container is considered to have started, and `failureThreshold` is used thereafter.

For example:

- The container fails 20 probes at startup.  It is considered to have failed, and is terminated.
- The container fails 10 probes at startup, succeeds once, and fails 3 more probes.  The probe is considered to have started at 100 seconds, and even though it has hit the failure condition of 3 consecutive failures 140 seconds into its lifetime (less than the 200 permitted by `maxInitialFailureCount`, it is considered to have failed and is terminated.

Reply:
Also, what happens if `maxInitialFailureCount` is set to less than `failureThreshold` (which needs to be addressed in the code as well)?

Reply:
I am going to add clear numerical examples...